TST: refactor a couple tests to avoid the use of random numbers (table)

neutrinoceros opened the pull request:
### Description

~Following @nstarman's suggestion from https://github.com/astropy/astropy/pull/16086/files#r1567914977
This is step 0 of #17026~

Update: the scope of this PR was changed to removing random-number generators in a couple simple test cases in the `table` test suite.

<!-- Optional opt-out -->

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

pllim commented:
Can we apply it to a small subpackage here just to see it used? Weird to add something new without it being used even if we know there will be follow-up. Thanks!

neutrinoceros commented:
Done. This is now applied to the table subpackage as it's one of the smallest changesets from #16086 (excluding packages that only get noqas)

pllim commented on astropy/table/table_helpers.py:
```diff
@@ -21,22 +21,21 @@ class TimingTables:
         self.table = Table(masked=self.masked)
 
         # Create column with mixed types
-        np.random.seed(12345)
```
Looks like with the new fixture, the ability to change seed is gone?

neutrinoceros replied:
Indeed. But if we're going to switch away from the legacy API we're losing the exact reproducibility of test cases anyway.
Quoting numpy docs:
> The RandomState provides access to legacy generators... This class should only be used if it is essential to have randoms that are identical to what would have been produced by previous versions of NumPy.

So, if tests still pass, it means we don't *need* exact reproducibility across this migration, but repeatability still seems important, which is why I *do* set a seed in the fixture.

pllim replied:
What if someone wants to run tests with different seeds on purpose for some reason?

neutrinoceros replied:
If we want to support that then I think we should *not* have a global fixture for it and we should instead keep using 
`np.random.default_rng` directly as Mridul did in #16086. We could have a parametrized fixture, but then we're back to square one and we don't gain anything over using `rng = np.random.default_rng(0)` directly.

I don't have strong opinions here, I'm honestly fine with how #16086 turned out *without* this fixture, I just didn't want @nstarman's review to go ignored.

pllim replied:
I don't see this fixture being necessary either. In saving some SLOC we lose flexibility. I'd rather have a bit more lines of code. I wonder what other maintainers think.

nstarman replied:
We can make the fixture reproducible!
See https://github.com/nengo/pytest-rng for an example where the seed can be set.

pllim replied:
So... based on discussions elsewhere, looks like only removing this `TimingTables` and the tiny refactor `test_item_access.py` are relevant now?

pllim commented on astropy/conftest.py:
```diff
@@ -63,6 +65,12 @@ def fast_thread_switching():
 
 
+@pytest.fixture
+def rng():
```
`rng` is also too cryptic for a fixture that would be used package-wide. Maybe something a bit more descriptive?

neutrinoceros replied:
how about `random_number_generator` ?

pllim replied:
If people think that is too long, `random_num_gen` might also work for me but I cannot speak for everyone.

nstarman replied:
IDK. Anyone using Astropy is using Numpy (/ an Array API library in future) and tHat's the standard name.

pllim replied:
I think we're leaning towards it is not necessary at all.

neutrinoceros commented on astropy/table/table_helpers.py:
```diff
@@ -21,22 +21,21 @@ class TimingTables:
     """
 
-    def __init__(self, size=1000, masked=False):
+    def __init__(self, rng, size=1000, masked=False):
```
Not sure I get what you mean here. The test seems to work as is and I don't see how `autouse=True` would help.

pllim replied:
OK. I had to use that in the past for some other fixture. If you don't need it, then nvm.

eerovaher replied:
This is not a test, so `rng` is not a `pytest` fixture here but a normal function parameter. The reason this backwards incompatible API change didn't cause any failures in CI is that the `TimingTables` class is not used anywhere. It would be better to remove it entirely instead of updating it.

neutrinoceros replied:
Ah, that explains it. Well, removing it now then.

eerovaher commented on astropy/table/tests/test_item_access.py:
```diff
@@ -242,13 +242,13 @@ def test_np_where(self, table_data):
 
-    def test_np_integers(self, table_data):
+    def test_np_integers(self, rng, table_data):
         """
         Select rows using numpy integers.  This is a regression test for a
         py 3.3 failure mode
         """
         t = table_data.Table(table_data.COLS)
-        idxs = np.random.randint(len(t), size=2)
+        idxs = rng.integers(len(t), size=2)
         t[idxs[1]]
```
The pull request that added the test says that in Python 2 `isinstance(np.int64(0), int)` was `True`, but in Python 3 it is `False`. This caused a bug with Python 3 where trying to access rows in a table using a `np.integer` didn't work. From the pull request description it is clear that the test is about the type of `idxs` and not its value. Generating `idxs` randomly is therefore not any better than hardcoding a value.

I should also point out that this test insists on generating two random values despite using only one (and needing zero). And the real problem with this test is that it doesn't assert anything. It just ensures that `t[idxs[1]]` doesn't raise an error, but it doesn't check if the output is a `Row` and not a `Column` or a `Table` or `None`.

neutrinoceros replied:
You're right. I'll patch this one to remove rng entierly.

neutrinoceros commented on astropy/table/tests/test_groups.py:
```diff
@@ -729,7 +729,7 @@ def test_group_mixins_unsupported(col):
 @pytest.mark.parametrize("add_index", [False, True])
-def test_group_stable_sort(add_index):
+def test_group_stable_sort(rng, add_index):
     """Test that group_by preserves the order of the table.
```
done !

eerovaher commented on astropy/table/tests/test_item_access.py:
```diff
@@ -248,8 +248,9 @@ def test_np_integers(self, table_data):
         t = table_data.Table(table_data.COLS)
-        idxs = np.random.randint(len(t), size=2)
-        t[idxs[1]]
+        row = t[np.int64(0)]
+        if not t.masked:
+            assert row == t[0]
```
If something goes wrong and instead of a `Row` we get a `Column` then this test would still succeed. Why not just
```suggestion
        assert type(t[np.int64(0)]) is Row
```
Or maybe there is some parametrized test somewhere already that checks what is the output of `Table` indexing based on the type of the index, and if there is then this test could be added as another parameter.

neutrinoceros replied:
I'm going to go with your suggestion as is

neutrinoceros commented:
Re: https://github.com/astropy/astropy/pull/17027#discussion_r1765779823

I rebased and cleaned-up this PR in this direction (note that I also kept the *other* small test refactor)

eerovaher commented on astropy/table/tests/test_item_access.py:
```diff
@@ -248,8 +250,8 @@ def test_np_integers(self, table_data):
         t = table_data.Table(table_data.COLS)
-        idxs = np.random.randint(len(t), size=2)
-        t[idxs[1]]
+        row = t[np.int64(0)]
```
```suggestion
```

pllim reviewed (approved):
Thanks! Let's get this in.
